### PR TITLE
Move state root to block header

### PR DESCRIPTION
**Pull request by fyrchik**
Prototype for #1498 .
Testing done:
1. 4-node privnet works as expected, stateroots are changed.

### CODE REVIEW

**Review by roman-khimov — CHANGES_REQUESTED** [review]
Looks better than expected, actually. Still, some modifications are needed:
 * more tests specific to stateroot-enabled chains and consensus (especially for verifyRequest and AddBlock)
 * block JSON format should contain stateroot data
 * we certainly need getproof/verifyproof RPC calls to be ported over here
 * I'm not sure about consensus recovery messages, looks like they don't need to be modified, but they were a problem on 2.x, so it'd be nice to check them again

File: pkg/consensus/payload.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ type (
 		ViewNumber byte
 
 		payload io.Serializable
+		// stateRootEnabled speifies if state root is exchanged during consensus.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/consensus/payload.go]: Fixed.

File: pkg/core/block/block_base.go [hunk 2]
```diff
@@ -61,6 +66,7 @@ type baseAux struct {
 	Timestamp     uint64                `json:"time"`
 	Index         uint32                `json:"index"`
 	NextConsensus string                `json:"nextconsensus"`
+	PrevStateRoot *util.Uint256         `json:"previousstateroot,omitempty"`
```
Comment (roman-khimov) [inline pkg/core/block/block_base.go:69]: It's `StateRoot` in the block and `previousstateroot` in JSON, we better use one name everywhere.

File: pkg/rpc/server/server.go [hunk 3]
```diff
@@ -774,6 +775,14 @@ func (s *Server) contractScriptHashFromParam(param *request.Param) (util.Uint160
 	return result, nil
 }
 
+func (s *Server) getStateHeight(_ request.Params) (interface{}, *response.Error) {
+	height := s.chain.BlockHeight()
+	return &result.StateHeight{
+		BlockHeight: height,
+		StateHeight: height,
```
Comment (roman-khimov) [inline pkg/rpc/server/server.go:848]: It's supposed to answer with verified state height, so it's `-1` and only if stateroot is enabled.

File: pkg/core/block/block_base.go (hunk 2 above)
Comment (fyrchik) [reply pkg/core/block/block_base.go:69]: Fixed.

Comment (fyrchik) [reply pkg/core/block/block_base.go:69]: `previous` in JSON to be unified with `previousblockhash`.

File: pkg/rpc/server/server.go (hunk 3 above)
Comment (fyrchik) [reply pkg/rpc/server/server.go]: Returned 0 as default.

File: pkg/rpc/server/server.go [hunk 4]
```diff
@@ -774,6 +775,18 @@ func (s *Server) contractScriptHashFromParam(param *request.Param) (util.Uint160
 	return result, nil
 }
 
+func (s *Server) getStateHeight(_ request.Params) (interface{}, *response.Error) {
+	var height = s.chain.BlockHeight()
+	var stateHeight uint32
+	if s.chain.GetConfig().StateRootInHeader {
+		stateHeight = height
```
Comment (roman-khimov) [inline pkg/rpc/server/server.go:848]: `-1`

Comment (roman-khimov) [reply pkg/rpc/server/server.go:856]: Still there.

Comment (fyrchik) [reply pkg/rpc/server/server.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**